File: 0020/stack.py
class Element:

    def __init__(self, val):
        self.val = val
        self.prev = None


class Stack:

    def __init__(self):
        self.head = None
        self.length = 0

    def push(self, val):
        element = Element(val)
        element.prev = self.head
        self.head = element
        self.length += 1

    def pop(self):
        if self.head is None:
            return None
        element = self.head
        self.head = self.head.prev
        self.length -= 1
        return element


class Solution:

    def isValid(self, s: str) -> bool:
        stack = Stack()
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.push(c)
            else:
                element = stack.pop()
                if element is None:
                    return False
                if c == ')' and element.val != '(':
                    return False
                elif c == '}' and element.val != '{':
                    return False
                elif c == ']' and element.val != '[':
                    return False
        return True if stack.length == 0 else False


if __name__ == '__main__':
    sol = Solution()
    assert sol.isValid('()') is True
    assert sol.isValid('()[]{}') is True
    assert sol.isValid('(]') is False
    assert sol.isValid('([)]') is False
    assert sol.isValid('{[]}') is True
    assert sol.isValid('}{') is False
    assert sol.isValid('()(') is False

File: 0020/stack.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <stack>
#include <string>

using namespace std;

class Solution {
 public:
  bool isValid(string s) {
    stack<char> st;
    for (const char& c : s) {
      if (c == '(' || c == '[' || c == '{') {
        st.push(c);
      } else {
        if (st.empty()) { return false; }
        if ((st.top() == '(' && c == ')') ||
            (st.top() == '[' && c == ']') ||
            (st.top() == '{' && c == '}')) {
          st.pop();
        } else {
          return false;
        }
      }
    }
    return st.empty();
  }
};

int main() {
  Solution sol;
  string s;
  bool expected;

  s = "()", expected = true;
  EXPECT_EQ(expected, sol.isValid(s));

  s = "()[]{}", expected = true;
  EXPECT_EQ(expected, sol.isValid(s));

  s = "(]", expected = false;
  EXPECT_EQ(expected, sol.isValid(s));

  s = "([)]", expected = false;
  EXPECT_EQ(expected, sol.isValid(s));

  s = "{[]}", expected = true;
  EXPECT_EQ(expected, sol.isValid(s));

  s = "[", expected = false;
  EXPECT_EQ(expected, sol.isValid(s));

  exit(EXIT_SUCCESS);
}
